modules/__tests__/transitionHooks-test.js

```diff
@@ -121,6 +136,35 @@ describe('When a router enters a branch', function () {
     })
   })
 
+  it('calls the route leave hooks when leaving the route', function (done) {
+    const history = createHistory('/news')
+
+    // Stub this function to exercise the code path.
+    history.listenBeforeUnload = () => (() => {})
+
+    render(<Router history={history} routes={routes}/>, node, function () {
+      expect(leaveHookSpy.calls.length).toEqual(0)
+      history.push('/inbox')
+      expect(leaveHookSpy.calls.length).toEqual(1)
+      history.push('/news')
+      expect(leaveHookSpy.calls.length).toEqual(1)
+      history.push('/inbox')
+      expect(leaveHookSpy.calls.length).toEqual(2)
+      done()
+    })
+  })
+
+  it('does not call removed route leave hooks', function (done) {
+    const history = createHistory('/news')
+
+    render(<Router history={history} routes={routes}/>, node, function () {
+      removeLeaveHook()
```

Comment:
The test above this one does - `<NewsFeed>` doesn't tear down its listener, and without the new logic, that test would fail on mismatching the counts.